### PR TITLE
QL: improve the "this block-comment should have been a QLDoc"-query

### PR DESCRIPTION
**Pull request by erik-krogh**
This would have caught a recent QLDoc check that I spent way too long trying to fix, where I didn't realise that the top-level comment was not a QLDoc.

I also fixed a whole bunch of the issues found by the query.   
Not all, some results are benign.  

And also some improvements to how we match up QLDoc in QL-for-QL.

### CODE REVIEW

**Review by michaelnebel — APPROVED** [review]
C# LGTM

File: ql/ql/src/queries/style/NonDocBlock.ql [hunk 1]
```diff
@@ -29,24 +29,59 @@ int getLineAboveNodeThatCouldHaveDoc(File file) {
...
-  canHaveQLDoc(node) and
+  (
+    canHaveQLDoc(node) and
+    comment = getCommentAtEnd(node.getLocation().getFile(), node.getLocation().getStartLine() - 1) and
+    nodeDescrip = "the below code"
+    or
+    node instanceof TopLevel and
+    comment = getCommentAtStart(node.getLocation().getFile(), 1) and
+    nodeDescrip = "the file"
+  ) and
```
Comment (geoffw0) [inline ql/ql/src/queries/style/NonDocBlock.ql:82]: It feels like we could remove some redundant logic here by adding a `reason` parameter to `getACommentThatCouldBeQLDoc` returning `"the below code"` or `"the file"`.

**Review by geoffw0 — APPROVED** [review]
C++ differences LGTM.

File: ql/ql/src/queries/style/NonDocBlock.ql (hunk 1 above)
Comment (erik-krogh) [reply ql/ql/src/queries/style/NonDocBlock.ql:82]: Yeah, that could save some duplication.  
But that sounds like something for a followup, this seems good enough to merge.

**Pull request merged**